course_UvA-DL/contrastive-learning/SimCLR.py

```diff
@@ -0,0 +1,865 @@
+# SimCLR thereby applies the InfoNCE loss, originally proposed by [Aaron van den Oord et al. ](https://arxiv.org/abs/1807.03748) for contrastive learning.
+# In short, the InfoNCE loss compares the similarity of $z_i$ and $z_j$ to the similarity of $z_i$ to any other representation in the batch by performing a softmax over the similarity values.
+# The loss can be formally written as:
+# $$
+# \ell_{i,j}=-\log \frac{\exp(\text{sim}(z_i,z_j)/\tau)}{\sum_{k=1}^{2N}\mathbb{1}_{[k\neq i]}\exp(\text{sim}(z_i,z_k)/\tau)}=-\text{sim}(z_i,z_j)/\tau+\log\left[\sum_{k=1}^{2N}\mathbb{1}_{[k\neq i]}\exp(\text{sim}(z_i,z_k)/\tau)\right]
+# $$
+# The function $\text{sim}$ is a similarity metric, and the hyperparameter $\tau$ is called temperature determining how peaked the distribution is.
+# Since many similarity metrics are bounded, the temperature parameter allows us to balance the influence of many dissimilar image patches versus one similar patch.
+# The similarity metric that is used in SimCLR is cosine similarity, as defined below:
+# $$
+# \text{sim}(z_i,z_j) = \frac{z_i^\top \cdot z_j}{||z_i||\cdot||z_j||}
+# $$
+# The maximum cosine similarity possible is $1$, while the minimum is $-1$.
+# In general, we will see that the features of two different images will converge to a cosine similarity around zero since the minimum, $-1$, would require $z_i$ and $z_j$ to be in the exact opposite direction in all feature dimensions, which does not allow for great flexibility.
+#
+# Finally, now that we have discussed all details, let's implement SimCLR below as a PyTorch Lightning module:
+
+
+# %%
+class SimCLR(pl.LightningModule):
+
+    def __init__(self, hidden_dim, lr, temperature, weight_decay, max_epochs=500):
+        super().__init__()
+        self.save_hyperparameters()
+        assert self.hparams.temperature > 0.0, 'The temperature must be a positive float!'
+        # Base model f(.)
+        self.convnet = torchvision.models.resnet18(
+            pretrained=False, num_classes=4 * hidden_dim
+        )  # num_classes is the output size of the last linear layer
+        # The MLP for g(.) consists of Linear->ReLU->Linear
+        self.convnet.fc = nn.Sequential(
+            self.convnet.fc,  # Linear(ResNet output, 4*hidden_dim)
+            nn.ReLU(inplace=True),
+            nn.Linear(4 * hidden_dim, hidden_dim)
+        )
+
+    def configure_optimizers(self):
+        optimizer = optim.AdamW(self.parameters(), lr=self.hparams.lr, weight_decay=self.hparams.weight_decay)
+        lr_scheduler = optim.lr_scheduler.CosineAnnealingLR(
+            optimizer, T_max=self.hparams.max_epochs, eta_min=self.hparams.lr / 50
+        )
+        return [optimizer], [lr_scheduler]
+
+    def info_nce_loss(self, batch, mode='train'):
```

Comment:
rather full name instead of `nce`

Reply:
Is `info_noise_contrastive_estimation_loss` more informative than `info_nce_loss` though? I think InfoNCE is more popular than the long version, but I am also ok with the long name.

Reply:
I would say so, but just my opinions (Python is talkative languge)